README.rst
Wikia
=========

**Wikia** is a Python library that makes it easy to access and parse
data from any `wikia website <https://www.wikia.com>`

Search a wikia, get article summaries, get data like links and images
from a page, and more. Wikipedia wraps the `Wikia API
<http://api.wikia.com/wiki/Wikia_API_Wiki>`__ so you can focus on using
wikia data, not getting it.



.. code:: python

  >>> import wikia
  >>> print wikia.summary("Runescape", "Banshee")
  # Banshees are Slayer monsters that require level 15 Slayer to kill. They frequently drop 13 noted pure essence, making them an alternative source of essence. Additionally, banshees tend to frequently drop many different types of herbs. Mighty banshees are a higher-levelled alternative, if this is given as your Slayer assignment.

  >>> wikia.search("Runescape", "Forest")
  # [u'Forest', u'Forester', u'Freaky Forester', u'Forester\'s Arms', u'Talking Forest', u'Jungle Forester', u'Dense forest', u'Forester hat', u'Ogre forester hat', u'Forester (Burgh de Rott Ramble)']

  >>> drakan = wikia.page("Runescape", "Castle Drakan")
  >>> drakan.title
  # u'Castle Drakan'
  >>> drakan.url
  # u'http://runescape.wikia.com/wiki/Castle_Drakan'
  >>> drakan.content
  # u'Castle Drakan is the home of Lord Drakan, the vampyre lord of Morytania. Found just north of Meiyerditch, it looms over the Sanguinesti region'...

  >>> wikia.set_lang("nl") # Dutch
  >>> wikia.summary("Runescape", "Runes", sentences=1)
  # Runes, of Magische runes zijn kleine gewichtloze steentjes waarmee spelers een spreuk kunnen uitvoeren.

Note: this library was designed for ease of use and simplicity, not for advanced use.

Installation
------------

To install Wikia, simply run:

::

  $ pip install wikia

Wikia is compatible with Python 2.6+ (2.7+ to run unittest discover) and Python 3.3+.

Documentation
-------------

Read the docs at http://api.wikia.com/wiki/Wikia_API_Wiki.

To run tests, clone the `respository on GitHub <https://github.com/timidger/Wikia>`__, then run:

::

  $ pip install -r requirements.txt
  $ bash runtests  # will run tests for python and python3
  $ python -m unittest discover wikia/tests/ '*test.py'  # manual style

in the root project directory.

To build the documentation yourself, after installing requirements.txt, run:

::

  $ pip install sphinx
  $ cd docs/
  $ make html

License
-------

MIT licensed. See the `LICENSE
file <https://github.com/Timidger/Wikiaa/blob/master/LICENSE>`__ for
full details.

Credits
-------

-  `wiki-api <https://github.com/richardasaurus/wiki-api>`__ by
   @richardasaurus for inspiration
-  @nmoroze and @themichaelyang for feedback and suggestions
-  The `Wikimedia
   Foundation <http://wikimediafoundation.org/wiki/Home>`__ for giving
   the world free access to data
-  @goldsmith for making such a fantastic library to fork
-  /u/captainmeta4 for giving the idea for a reddit bot to post game wiki info
   like auto-wiki bot



CHANGELOG.md
# Changelog

### Version 1.4.4
* Fixed a bug with non-existant references and sections property being accessed

## Current
### Version 1.4.3
* Fixed a bug with the images, where some of them would not be retrieved properly
### Version 1.4.2
* Fixed a bug with images, where JPG would not be properly formatted
* Removed redirect/missing page code checks, because they might not be needed

### Version 1.4.1
* Expanded image extraction to include all images in the article

### Version 1.4
* Finished conversion to the wikia API from the wikipedia API

### Version 1.3.1

* Determine package version without importing ``wikipedia`` in setup.py. Fixes [#50](https://github.com/goldsmith/Wikipedia/issues/50) reported by [@arcolife](https://github.com/arcolife)

* Update ``requests`` dependency to v2.3.0

## Version 1.3

* wikipedia.languages() for easy access to all language prefixes
* Conditional check for normalization in redirect queries. Fixes [#47](https://github.com/goldsmith/Wikipedia/issues/47)
* Remove pip requirement to fix pip installation error. Fixes [#46](https://github.com/goldsmith/Wikipedia/issues/46#issuecomment-44221725) reported by [@oquidav](https://github.com/oquidave)

### Version 1.2.1

* Refactor query functions to standardize & fix functionality of WikipediaPage properties that return a list. Fixes [#38](https://github.com/goldsmith/Wikipedia/issues/38) reported by [@gwezerek](https://github.com/gwezerek)

* Use official Mediawiki API 'redirects' key to avoid redirection parse errors. Fixes [#28](https://github.com/goldsmith/Wikipedia/issues/28)
  reported by [@shichao-an](https://github.com/shichao-an) and [#32](https://github.com/goldsmith/Wikipedia/issues/32) reported by [@dmirylenka](https://github.com/dmirylenka)

## Version 1.2

* Add revision_id and parent_id properties. PR [#23](https://github.com/goldsmith/Wikipedia/pull/23) by [@fusiongyro](https://github.com/fusiongyro)
* Add changeable User-Agent. PR [#33](https://github.com/goldsmith/Wikipedia/pull/33) by [@crazybmanp](https://github.com/crazybmanp)
* Add geosearch functionality. PR [#40](https://github.com/goldsmith/Wikipedia/pull/40) by [@Kazuar](https://github.com/Kazuar)

## Version 1.1

* Add limited ability to access section titles on a page.  PR [#18](https://github.com/goldsmith/Wikipedia/pull/18) by [@astavonin](https://github.com/astavonin)
* Add optional rate limiting. Closes issue [#20](https://github.com/goldsmith/Wikipedia/pull/20) reported by [@mobeets](https://github.com/mobeets)
* Add HTTPTimeout exception

### Version 1.0.2

* Fix installation issue on some Python 3 machines. Closes issue [#15](https://github.com/goldsmith/Wikipedia/issues/15) by [@wronglink](https://github.com/wronglink)
* Add Python 3 support on PyPI

## Version 1.0

* Add international support
* Fix continue values with Mediawiki API
* Support Python 3

### Version 0.9

* Initial functionality
* Add documentation and upload to ReadTheDocs


tests/lang_test.py
# -*- coding: utf-8 -*-
import unittest

from wikia import wikia


class TestLang(unittest.TestCase):
  """Test the ability for wikia to change the language of the API being accessed."""

  def test_lang(self):
    wikia.set_lang("fr")
    self.assertEqual(wikia.API_URL, 'http://fr.www.wikia.com/api/v1')
    wikia.set_subwikia("runescape")
    self.assertEqual(wikia.API_URL, 'http://fr.runescape.wikia.com/api/v1')


tests/page_test.py
# -*- coding: utf-8 -*-
from decimal import Decimal
import unittest

from wikia import wikia
from request_mock_data import mock_data


# mock out _wiki_request
def _wiki_request(params):
  return mock_data["_wiki_request calls"][tuple(sorted(params.items()))]
wikia._wiki_request = _wiki_request


class TestPageSetUp(unittest.TestCase):
  """Test the functionality of wikia.page's __init__ and load functions."""

  def test_missing(self):
    """Test that page raises a PageError for a nonexistant page."""
    # Callicarpa?
    purpleberry = lambda: wikia.page("purpleberry")
    self.assertRaises(wikia.PageError, purpleberry)

  def test_redirect_true(self):
    """Test that a page successfully redirects a query."""
    # no error should be raised if redirect is test_redirect_true
    mp = wikia.page("Menlo Park, New Jersey")

    self.assertEqual(mp.title, "Edison, New Jersey")
    self.assertEqual(mp.url, "http://en.wikia.org/wiki/Edison,_New_Jersey")

  def test_redirect_false(self):
    """Test that page raises an error on a redirect when redirect == False."""
    mp = lambda: wikia.page("Menlo Park, New Jersey", redirect=False)
    self.assertRaises(wikia.RedirectError, mp)

  def test_redirect_no_normalization(self):
    """Test that a page with redirects but no normalization query loads correctly"""
    the_party = wikia.page("Communist Party")
    self.assertIsInstance(the_party, wikia.WikiaPage)
    self.assertEqual(the_party.title, "Communist party")

  def test_redirect_with_normalization(self):
    """Test that a page redirect with a normalized query loads correctly"""
    the_party = wikia.page("communist Party")
    self.assertIsInstance(the_party, wikia.WikiaPage)
    self.assertEqual(the_party.title, "Communist party")

  def test_redirect_normalization(self):
    """Test that a page redirect loads correctly with or without a query normalization"""
    capital_party = wikia.page("Communist Party")
    lower_party = wikia.page("communist Party")

    self.assertIsInstance(capital_party, wikia.WikiaPage)
    self.assertIsInstance(lower_party, wikia.WikiaPage)
    self.assertEqual(capital_party.title, "Communist party")
    self.assertEqual(capital_party, lower_party)

class TestPage(unittest.TestCase):
  """Test the functionality of the rest of wikia.page."""

  def setUp(self):
    # shortest wikia articles with images and sections
    self.celtuce = wikia.page("Celtuce")
    self.cyclone = wikia.page("Tropical Depression Ten (2005)")
    self.great_wall_of_china = wikia.page("Great Wall of China")

  def test_from_page_id(self):
    """Test loading from a page id"""
    self.assertEqual(self.celtuce, wikia.page(pageid=1868108))

  def test_title(self):
    """Test the title."""
    self.assertEqual(self.celtuce.title, "Celtuce")
    self.assertEqual(self.cyclone.title, "Tropical Depression Ten (2005)")

  def test_url(self):
    """Test the url."""
    self.assertEqual(self.celtuce.url, "http://en.wikia.org/wiki/Celtuce")
    self.assertEqual(self.cyclone.url, "http://en.wikia.org/wiki/Tropical_Depression_Ten_(2005)")

  def test_content(self):
    """Test the plain text content."""
    self.assertEqual(self.celtuce.content, mock_data['data']["celtuce.content"])
    self.assertEqual(self.cyclone.content, mock_data['data']["cyclone.content"])

  def test_revision_id(self):
    """Test the revision id."""
    self.assertEqual(self.celtuce.revision_id, mock_data['data']["celtuce.revid"])
    self.assertEqual(self.cyclone.revision_id, mock_data['data']["cyclone.revid"])

  def test_parent_id(self):
    """Test the parent id."""
    self.assertEqual(self.celtuce.parent_id, mock_data['data']["celtuce.parentid"])
    self.assertEqual(self.cyclone.parent_id, mock_data['data']["cyclone.parentid"])


  def test_summary(self):
    """Test the summary."""
    self.assertEqual(self.celtuce.summary, mock_data['data']["celtuce.summary"])
    self.assertEqual(self.cyclone.summary, mock_data['data']["cyclone.summary"])

  def test_images(self):
    """Test the list of image URLs."""
    self.assertEqual(sorted(self.celtuce.images), mock_data['data']["celtuce.images"])
    self.assertEqual(sorted(self.cyclone.images), mock_data['data']["cyclone.images"])

  def test_references(self):
    """Test the list of reference URLs."""
    self.assertEqual(self.celtuce.references, mock_data['data']["celtuce.references"])
    self.assertEqual(self.cyclone.references, mock_data['data']["cyclone.references"])

  def test_links(self):
    """Test the list of titles of links to Wikia pages."""
    self.assertEqual(self.celtuce.links, mock_data['data']["celtuce.links"])
    self.assertEqual(self.cyclone.links, mock_data['data']["cyclone.links"])

  def test_html(self):
    """Test the full HTML method."""
    self.assertEqual(self.celtuce.html(), mock_data['data']["celtuce.html"])

  def test_sections(self):
    """Test the list of section titles."""
    self.assertEqual(sorted(self.cyclone.sections), mock_data['data']["cyclone.sections"])

  def test_section(self):
    """Test text content of a single section."""
    self.assertEqual(self.cyclone.section("Impact"), mock_data['data']["cyclone.section.impact"])
    self.assertEqual(self.cyclone.section("History"), None)

  def test_coordinates(self):
    """Test geo coordinates of a page"""
    lat, lon = self.great_wall_of_china.coordinates
    self.assertEqual(str(lat.quantize(Decimal('1.000'))), mock_data['data']['great_wall_of_china.coordinates.lat'])
    self.assertEqual(str(lon.quantize(Decimal('1.000'))), mock_data['data']['great_wall_of_china.coordinates.lon'])


tests/search_test.py
# -*- coding: utf-8 -*-
import unittest

from collections import defaultdict

from wikia import wikia
from request_mock_data import mock_data


# mock out _wiki_request
class _wiki_request(object):

  calls = defaultdict(int)

  @classmethod
  def __call__(cls, params):
    cls.calls[params.__str__()] += 1
    return mock_data["_wiki_request calls"][tuple(sorted(params.items()))]

wikia._wiki_request = _wiki_request()


class TestSearch(unittest.TestCase):
  """Test the functionality of wikia.search."""

  def test_search(self):
    """Test parsing a Wikipedia request result."""
    self.assertEqual(wikia.search("Barack Obama"), mock_data['data']["barack.search"])

  def test_limit(self):
    """Test limiting a request results."""
    self.assertEqual(wikia.search("Porsche", results=3), mock_data['data']["porsche.search"])


wikipedia/wikipedia.py
from __future__ import unicode_literals

import requests
import time
from bs4 import BeautifulSoup
from datetime import datetime, timedelta
from decimal import Decimal

from .exceptions import (
  PageError, DisambiguationError, RedirectError, HTTPTimeoutError,
  WikipediaException, ODD_ERROR_MESSAGE)
from .util import cache, stdout_encode, debug

API_URL = "http://{site}{lang}.gamepedia.com/api.php?"
RATE_LIMIT = False
RATE_LIMIT_MIN_WAIT = None
RATE_LIMIT_LAST_CALL = None
USER_AGENT = 'gamepedia (https://github.com/timidger/Gamepedia/)'


## Change how this API works
# Remove the global functions
# Make this have to operate through an object


def set_user_agent(user_agent_string):
  '''
  Set the User-Agent string to be used for all requests.

  Arguments:

  * user_agent_string - (string) a string specifying the User-Agent header
  '''
  global USER_AGENT
  USER_AGENT = user_agent_string


def set_rate_limiting(rate_limit, min_wait=timedelta(milliseconds=50)):
  '''
  Enable or disable rate limiting on requests to the Mediawiki servers.
  If rate limiting is not enabled, under some circumstances (depending on
  load on Wikipedia, the number of requests you and other `wikipedia` users
  are making, and other factors), Wikipedia may return an HTTP timeout error.

  Enabling rate limiting generally prevents that issue, but please note that
  HTTPTimeoutError still might be raised.

  Arguments:

  * rate_limit - (Boolean) whether to enable rate limiting or not

  Keyword arguments:

  * min_wait - if rate limiting is enabled, `min_wait` is a timedelta describing the minimum time to wait before requests.
         Defaults to timedelta(milliseconds=50)
  '''
  global RATE_LIMIT
  global RATE_LIMIT_MIN_WAIT
  global RATE_LIMIT_LAST_CALL

  RATE_LIMIT = rate_limit
  if not rate_limit:
    RATE_LIMIT_MIN_WAIT = None
  else:
    RATE_LIMIT_MIN_WAIT = min_wait

  RATE_LIMIT_LAST_CALL = None


@cache
def search(site, query, lang=None, results=10, suggestion=False):
  '''
  Do a search on the gamepedia site for `query`.

  Keyword arguments:

  * lang - The language the Gamepedia article is written in
  * results - the maxmimum number of results returned
  * suggestion - if True, return results and suggestion (if any) in a tuple
  '''
  api_url = format_url(site, lang)
  search_params = {
    'list': 'search',
    'srprop': '',
    'srlimit': results,
    'limit': results,
    'srsearch': query,
    'site': site,
  }
  if suggestion:
    search_params['srinfo'] = 'suggestion'

  raw_results = _wiki_request(search_params, api_url)

  if 'error' in raw_results:
    if raw_results['error']['info'] in ('HTTP request timed out.', 'Pool queue is full'):
      raise HTTPTimeoutError(query)
    else:
      raise WikipediaException(raw_results['error']['info'])

  search_results = (d['title'] for d in raw_results['query']['search'])

  if suggestion:
    if raw_results['query'].get('searchinfo'):
      return list(search_results), raw_results['query']['searchinfo']['suggestion']
    else:
      return list(search_results), None

  return list(search_results)


@cache
def suggest(site, query, lang=None):
  '''
  Get a search suggestion for `query` on gamepedia site `site`
  in a certain language 'lang' (which should be a prefix).
  Returns a string or None if no suggestion was found.
  '''

  api_url = format_url(site, lang)
  search_params = {
    'list': 'search',
    'srinfo': 'suggestion',
    'srprop': '',
    'site': site,
  }
  search_params['srsearch'] = query

  raw_result = _wiki_request(search_params, api_url)

  if raw_result['query'].get('searchinfo'):
    return raw_result['query']['searchinfo']['suggestion']

  return None


def random(site, pages=1, lang=None):
  '''
  Get a list of random article titles from the gamepedia `site` in `lang`.

  .. note:: Random only gets articles from namespace 0, meaning no Category,
            User talk, or other meta-Wikipedia pages.

  Keyword arguments:

  * pages - the number of random pages returned (max of 10)
  * lang - The prefix for the language the site to be in
  '''
  api_url = format_url(site, lang)
  #http://en.wikipedia.org/w/api.php?action=query&list=random&rnlimit=5000&format=jsonfm
  query_params = {
    'list': 'random',
    'rnnamespace': 0,
    'rnlimit': pages,
  }

  request = _wiki_request(query_params, api_url)
  titles = [page['title'] for page in request['query']['random']]

  if len(titles) == 1:
    return titles[0]

  return titles


@cache
def summary(site, title, lang=None, sentences=0, chars=0,
           auto_suggest=True, redirect=True):
  '''
  Plain text summary of the page on the gamepedia site `site`.

  .. note:: This is a convenience wrapper - auto_suggest and redirect are enabled by default

  Keyword arguments:

  * sentences - if set, return the first `sentences` sentences (can be no greater than 10).
  * chars - if set, return only the first `chars` characters (actual text returned may be slightly longer).
  * auto_suggest - let Wikipedia find a valid page title for the query
  * redirect - allow redirection without raising RedirectError
  * lang - The prefix for the language the site to be in
  '''

  api_url = format_url(site, lang)
  # use auto_suggest and redirect to get the correct article
  # also, use page's error checking to raise DisambiguationError if necessary
  page_info = page(site, title, auto_suggest=auto_suggest, redirect=redirect)
  title = page_info.title
  pageid = page_info.pageid

  query_params = {
    'prop': 'revisions',
    'rvprop': 'content',
    'titles': title,
  }


  request = _wiki_request(query_params, api_url)
  summary = request['query']['pages'][pageid]['revisions'][0]["*"]

  return summary


def page(site, title=None, pageid=None, lang=None, auto_suggest=True,
        redirect=True, preload=False):
  '''
  Get a WikipediaPage object for the page with title `title` or the pageid
  `pageid` (mutually exclusive) from the gamepedia site.

  Keyword arguments:

  * site - The gamepedia site the article is on
  * title - the title of the page to load
  * pageid - the numeric pageid of the page to load
  * auto_suggest - let Wikipedia find a valid page title for the query
  * redirect - allow redirection without raising RedirectError
  * preload - load content, summary, images, references, and links during initialization
  '''

  if title is not None:
    if auto_suggest:
      results, suggestion = search(site, title, results=1, suggestion=True)
      try:
        title = suggestion or results[0]
      except IndexError:
        # if there is no suggestion or search results, the page doesn't exist
        raise PageError(title)
    return WikipediaPage(site, title, lang=lang, redirect=redirect, preload=preload)
  elif pageid is not None:
    return WikipediaPage(site, pageid=pageid, lang=lang, preload=preload)
  else:
    raise ValueError("Either a title or a pageid must be specified")



class WikipediaPage(object):
  '''
  Contains data from a Wikipedia page.
  Uses property methods to filter data from the raw HTML.
  '''

  def __init__(self, site, title=None, pageid=None, lang=None, redirect=True,
              preload=False, original_title=''):
    self.site = site
    if lang is not None:
      self.lang = lang
    if title is not None:
      self.title = title
      self.original_title = original_title or title
    elif pageid is not None:
      self.pageid = pageid
    else:
      raise ValueError("Either a title or a pageid must be specified")

    self.__load(redirect=redirect, preload=preload)

    if preload:
      for prop in ('content', 'summary', 'images', 'references', 'links', 'sections'):
        getattr(self, prop)

  def __repr__(self):
    return stdout_encode(u'<WikipediaPage \'{}\'>'.format(self.title))

  def __eq__(self, other):
    try:
      return (
        self.pageid == other.pageid
        and self.title == other.title
        and self.url == other.url
      )
    except:
      return False

  def __load(self, redirect=True, preload=False):
    '''
    Load basic information from Wikipedia.
    Confirm that page exists and is not a disambiguation/redirect.

    Does not need to be called manually, should be called automatically during __init__.
    '''
    query_params = {
      'prop': 'info|pageprops',
      'inprop': 'url',
      'ppprop': 'disambiguation',
      'redirects': '',
    }
    if not getattr(self, 'pageid', None):
      query_params['titles'] = self.title
    else:
      query_params['pageids'] = self.pageid

    request = self._wiki_request(query_params)

    query = request['query']
    pageid = list(query['pages'].keys())[0]
    page = query['pages'][pageid]

    # missing is present if the page is missing
    if 'missing' in page:
      if hasattr(self, 'title'):
        raise PageError(self.title)
      else:
        raise PageError(pageid=self.pageid)

    # same thing for redirect, except it shows up in query instead of page for
    # whatever silly reason
    elif 'redirects' in query:
      if redirect:
        redirects = query['redirects'][0]

        if 'normalized' in query:
          normalized = query['normalized'][0]
          assert normalized['from'] == self.title, ODD_ERROR_MESSAGE

          from_title = normalized['to']

        else:
          from_title = self.title

        assert redirects['from'] == from_title, ODD_ERROR_MESSAGE

        # change the title and reload the whole object
        self.__init__(redirects['to'], redirect=redirect, preload=preload)

      else:
        raise RedirectError(getattr(self, 'title', page['title']))

    # since we only asked for disambiguation in ppprop,
    # if a pageprop is returned,
    # then the page must be a disambiguation page
    elif 'pageprops' in page:
      query_params = {
        'prop': 'revisions',
        'rvprop': 'content',
        'rvparse': '',
        'rvlimit': 1
      }
      if hasattr(self, 'pageid'):
        query_params['pageids'] = self.pageid
      else:
        query_params['titles'] = self.title
      request = self._wiki_request(query_params)
      html = request['query']['pages'][pageid]['revisions'][0]['*']

      lis = BeautifulSoup(html).find_all('li')
      filtered_lis = [li for li in lis if not 'tocsection' in ''.join(li.get('class', []))]
      may_refer_to = [li.a.get_text() for li in filtered_lis if li.a]

      raise DisambiguationError(getattr(self, 'title', page['title']), may_refer_to)

    else:
      self.pageid = pageid
      self.title = page['title']
      self.url = page['fullurl']

  def __continued_query(self, query_params):
    '''
    Based on https://www.mediawiki.org/wiki/API:Query#Continuing_queries
    '''
    query_params.update(self.__title_query_param)

    last_continue = {}
    prop = query_params.get('prop', None)

    while True:
      params = query_params.copy()
      params.update(last_continue)

      request = self._wiki_request(params)

      if 'query' not in request:
        break

      pages = request['query']['pages']
      if 'generator' in query_params:
        for datum in pages.values():  # in python 3.3+: "yield from pages.values()"
          yield datum
      else:
        for datum in pages[self.pageid][prop]:
          yield datum

      if 'continue' not in request:
        break

      last_continue = request['continue']

  def _wiki_request(self, params):
    '''
    Wrapper for _wiki_request, automatically sends the correct api_url
    for the object's request
    '''
    return _wiki_request(params, self.api_url)

  @property
  def __title_query_param(self):
    if getattr(self, 'title', None) is not None:
      return {'titles': self.title, 'page' : self.title}
    else:
      return {'pageids': self.pageid}

  def html(self):
    '''
    Get full page HTML.

    .. warning:: This can get pretty slow on long pages.
    '''

    if not getattr(self, '_html', False):
      query_params = {
        'prop': 'revisions',
        'rvprop': 'content',
        'rvlimit': 1,
        'rvparse': '',
        'titles': self.title
      }

      request = self._wiki_request(query_params)
      self._html = request['query']['pages'][self.pageid]['revisions'][0]['*']

    return self._html

  @property
  def lang(self):
    '''
    The language prefix that the API uses requests the article be in a certain
    language. The prefix must be in the following list:
    <http://meta.wikimedia.org/wiki/List_of_Wikipedias>`_.
    '''
    if not getattr(self, '_lang', False):
        return None
    return self._lang

  @lang.setter
  def lang(self, lang):
    '''
    Change the language of the API being requested.
    Set `prefix` to one of the two letter prefixes found on the
    `list of all Wikipedias <http://meta.wikimedia.org/wiki/List_of_Wikipedias>`_.

    After setting the language, the cache for ``search``, ``suggest``, and ``summary`` will be cleared.

    .. note:: Make sure you search for page titles in the language that you have set.
    '''
    self._lang = lang
    for cached_func in (search, suggest, summary):
      cached_func.clear_cache()

  @property
  def site(self):
    '''
    The particular gamepedia we are requesting the data from.
    The site must be one of the following wikis: <http://www.gamepedia.com/wikis>
    '''
    return self._site

  @site.setter
  def site(self, site):
    '''
    Sets what gamepedia site the API will gather data from.
    Must be in the following list: <http://www.gamepedia.com/wikis>
    '''
    self._site = site

  @property
  def api_url(self):
    return format_url(self.site, self.lang)

  @property
  def content(self):
    '''
    Plain text content of the page, excluding images, tables, and other data.
    '''

    if not getattr(self, '_content', False):
      query_params = {
        'action': 'parse',
        'prop': 'wikitext',
      }
      if not getattr(self, 'title', None) is None:
         query_params['page'] = self.title
      else:
         query_params['pageid'] = self.pageid
      request = self._wiki_request(query_params)
      self._content     = request['parse']['wikitext']['*']
      #self._revision_id = request['query']['pages'][self.pageid]['revisions'][0]['revid']
      #self._parent_id   = request['query']['pages'][self.pageid]['revisions'][0]['parentid']

    return self._content

  @property
  def revision_id(self):
    '''
    Revision ID of the page.

    The revision ID is a number that uniquely identifies the current
    version of the page. It can be used to create the permalink or for
    other direct API calls. See `Help:Page history
    <http://en.wikipedia.org/wiki/Wikipedia:Revision>`_ for more
    information.
    '''

    if not getattr(self, '_revid', False):
      # fetch the content (side effect is loading the revid)
      self.content

    return self._revision_id

  @property
  def parent_id(self):
    '''
    Revision ID of the parent version of the current revision of this
    page. See ``revision_id`` for more information.
    '''

    if not getattr(self, '_parentid', False):
      # fetch the content (side effect is loading the revid)
      self.content

    return self._parent_id

  @property
  def summary(self):
    '''
    Plain text summary of the page.
    '''

    if not getattr(self, '_summary', False):
      query_params = {
        'prop': 'revisions',
        'rvprop': 'content',
      }
      if not getattr(self, 'title', None) is None:
         query_params['titles'] = self.title
      else:
         query_params['pageids'] = self.pageid

      request = self._wiki_request(query_params)
      self._summary = request['query']['pages'][self.pageid]['revisions'][0]["*"]
    return self._summary

  @property
  def images(self):
    '''
    List of URLs of images on the page.
    '''

    if not getattr(self, '_images', False):
      self._images = [
        page['imageinfo'][0]['url']
        for page in self.__continued_query({
          'generator': 'images',
          'gimlimit': 'max',
          'prop': 'imageinfo',
          'iiprop': 'url',
        })
        if 'imageinfo' in page
      ]

    return self._images

  @property
  def coordinates(self):
    '''
    Tuple of Decimals in the form of (lat, lon) or None
    '''
    if not getattr(self, '_coordinates', False):
      query_params = {
        'prop': 'coordinates',
        'colimit': 'max',
        'titles': self.title,
      }

      request = self._wiki_request(query_params)

      if 'query' in request:
        coordinates = request['query']['pages'][self.pageid]['coordinates']
        self._coordinates = (Decimal(coordinates[0]['lat']), Decimal(coordinates[0]['lon']))
      else:
        self._coordinates = None

    return self._coordinates

  @property
  def references(self):
    '''
    List of URLs of external links on a page.
    May include external links within page that aren't technically cited anywhere.
    '''

    if not getattr(self, '_references', False):
      def add_protocol(url):
        return url if url.startswith('http') else 'http:' + url

      self._references = [
        add_protocol(link['*'])
        for link in self.__continued_query({
          'prop': 'extlinks',
          'ellimit': 'max'
        })
      ]

    return self._references

  @property
  def links(self):
    '''
    List of titles of Wikipedia page links on a page.

    .. note:: Only includes articles from namespace 0, meaning no Category, User talk, or other meta-Wikipedia pages.
    '''

    if not getattr(self, '_links', False):
      self._links = [
        link['title']
        for link in self.__continued_query({
          'prop': 'links',
          'plnamespace': 0,
          'pllimit': 'max'
        })
      ]

    return self._links

  @property
  def sections(self):
    '''
    List of section titles from the table of contents on the page.
    '''

    if not getattr(self, '_sections', False):
      query_params = {
        'action': 'parse',
        'prop': 'sections',
      }
      query_params.update(self.__title_query_param)

      request = self._wiki_request(query_params)
      self._sections = [section['line'] for section in request['parse']['sections']]

    return self._sections

  def section(self, section_title):
    '''
    Get the plain text content of a section from `self.sections`.
    Returns None if `section_title` isn't found, otherwise returns a whitespace stripped string.

    This is a convenience method that wraps self.content.

    .. warning:: Calling `section` on a section that has subheadings will NOT return
           the full text of all of the subsections. It only gets the text between
           `section_title` and the next subheading, which is often empty.
    '''

    section = u"== {} ==".format(section_title)
    try:
      index = self.content.index(section) + len(section)
    except ValueError:
      return None

    try:
      next_index = self.content.index("==", index)
    except ValueError:
      next_index = len(self.content)

    return self.content[index:next_index].lstrip("=").strip()


@cache
def languages():
  '''
  List all the currently supported language prefixes (usually ISO language code).

  Can be inputted to `set_lang` to change the Mediawiki that `wikipedia` requests
  results from.

  Returns: dict of <prefix>: <local_lang_name> pairs. To get just a list of prefixes,
  use `wikipedia.languages().keys()`.
  '''
  response = self._wiki_request({
    'site': '',
    'meta': 'siteinfo',
    'siprop': 'languages'
  })

  languages = response['query']['languages']

  return {
    lang['code']: lang['*']
    for lang in languages
  }


def donate():
  '''
  Open up the Wikimedia donate page in your favorite browser.
  '''
  import webbrowser

  webbrowser.open('https://donate.wikimedia.org/w/index.php?title=Special:FundraiserLandingPage', new=2)


def format_url(site, lang=None):
  '''
  Formats the API url to point to a certain Gamepedia site. If a lang prefix
  is given, then that is used as the language.
  '''
  global API_URL
  return API_URL.format(site=site, lang="-" + lang if lang else "")


def _wiki_request(params, api_url):
  '''
  Make a request to the Gamepedia API using the given search parameters.
  Returns a parsed dict of the JSON response.
  '''
  global RATE_LIMIT_LAST_CALL
  global USER_AGENT

  params['format'] = 'json'
  if not 'action' in params:
    params['action'] = 'query'

  headers = {
    'User-Agent': USER_AGENT
  }

  if RATE_LIMIT and RATE_LIMIT_LAST_CALL and \
    RATE_LIMIT_LAST_CALL + RATE_LIMIT_MIN_WAIT > datetime.now():

    # it hasn't been long enough since the last API call
    # so wait until we're in the clear to make the request

    wait_time = (RATE_LIMIT_LAST_CALL + RATE_LIMIT_MIN_WAIT) - datetime.now()
    time.sleep(int(wait_time.total_seconds()))

  r = requests.get(api_url, params=params, headers=headers)

  if RATE_LIMIT:
    RATE_LIMIT_LAST_CALL = datetime.now()
  return r.json()
